lib/pbench/server/database/models/audit.py

```diff
@@ -0,0 +1,370 @@
+    SQLAlchemy errors reported through Audit operations.
+
+    The exception will identify the operation being performed and the config
+    key; the cause will specify the original SQLAlchemy exception.
+    """
+
+    def __init__(self, operation: str, params: JSONOBJECT, cause: str):
+        self.operation = operation
+        self.params = params
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Error {self.operation} {self.params!r}: {self.cause}"
+
+
+class AuditDuplicate(AuditError):
+    """
+    Attempt to commit a duplicate ServerConfig.
+    """
+
+    def __init__(self, name: str, cause: str):
+        self.name = name
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Duplicate config setting {self.name!r}: {self.cause}"
+
+
+class AuditNullKey(AuditError):
+    """
+    Attempt to commit a ServerConfig with an empty key.
+    """
+
+    def __init__(self, name: str, cause: str):
+        self.name = name
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Missing key value in {self.name!r}: {self.cause}"
+
+
+class AuditType(enum.Enum):
+    """
+    The type of Pbench Server resource affected by a CREATE, UPDATE, or DELETE
+    operation.
+    """
+
+    """Operation on a Dataset resource. This should include both the
+    object_id and object_name."""
+    DATASET = enum.auto()
+
+    """Operation on a system configuration setting. There's no meaningful id
+    or name; the 'attributes' column will record the actual attributes we
+    set."""
+    CONFIG = enum.auto()
+
+    """A default when no resource is affected"""
+    NONE = enum.auto()
+
+    """An Elasticsearch template. The base template name will be recorded in
+    object_name."""
+    TEMPLATE = enum.auto()
+
+    """Operation on an API token. There's no meaningful id or name."""
+    TOKEN = enum.auto()
+
+
+class AuditStatus(enum.Enum):
+
+    """Pending operation: this signals the beginning of an optation that might
+    fail or require further context later."""
+
+    BEGIN = enum.auto()
+
+    """Successful operation: an operation completed without problems."""
+    SUCCESS = enum.auto()
+
+    """Failed operation: an operation failed for reasons categorized by
+    AuditReason and possibly further detailed by a message."""
+    FAILURE = enum.auto()
+
+    """Warning: the operation didn't fail, but wasn't able to complete with
+    total success."""
+    WARNING = enum.auto()
+
+
+class AuditReason(enum.Enum):
+
+    """Permission denied."""
+
+    PERMISSION = enum.auto()
+
+    """Internal failure."""
+    INTERNAL = enum.auto()
+
+    """Consistency failure. For example, a corrupted tarball upload."""
+    CONSISTENCY = enum.auto()
+
+
+class Audit(Database.Base):
+    """
+    A framework to store Pbench audit records. These will track server
+    configuration changes as well as every mutation of user-visible data and
+    attribute each to a specific user and operation.
```

Comment:
As specified above, some operations don't have an ID or user, but as proposed earlier, we might want to attribute them to the unique ID of a server entity executing them.